src/syntax/instrumentPpx.ml

```diff
@@ -354,26 +340,43 @@ class instrumenter = object (self)
           l
         in
           Str.value ~loc rec_flag l
-    | Pstr_eval (e, a) ->
+    | Pstr_eval (e, a) when not (attribute_guard || extension_guard) ->
         Str.eval ~loc (wrap_expr Common.Toplevel_expr (self#expr e))
     | _ ->
         super#structure_item si
 
-  (*  Initializes storage and applies requested marks. *)
+  (* Guard these because they can carry payloads that we
+     do not want to instrument. *)
+  method! extension e =
+    extension_guard <- true;
+    let r = super#extension e in
+    extension_guard <- false;
+    r
+
+  method! attribute a =
+    attribute_guard <- true;
+    let r = super#attribute a in
+    attribute_guard <- false;
+    r
+
+  (* Initializes storage and applies requested marks. *)
   method! structure ast =
-    let ast = super#structure ast in
-    match get_filename ast with
-    | None -> ast
-    | Some file ->
-      if not (InstrumentState.is_file file) then
-        let header =
-          match !InstrumentArgs.mode with
-          | InstrumentArgs.Safe   -> safe file
-          | InstrumentArgs.Fast
-          | InstrumentArgs.Faster -> faster file
-        in
-        header :: ast
-      else
-        ast
+    (*let ts = String.concat "," (List.map typoo ast) in *)
+    if extension_guard || attribute_guard then
```

Comment:
I think this was necessary because there were two points that triggered instrumentation.

But you may also be right that it is no longer needed. I advise that you get rid of it and see if anything breaks, as we wrote tests for this. It would be faster to figure out if this line has meaning that way than me trying to recall 2 year old details.